sunpy/data/manager/manager.py

```diff
@@ -0,0 +1,135 @@
+import pathlib
+import functools
+from contextlib import contextmanager
+
+from sunpy.util.util import hash_file
+
+
+class DataManager:
+    """
+    DataManager
+    """
+
+    def __init__(self, cache):
+        # TODO: Folder prefix should be init argument
+        self._cache = cache
+
+        self._file_cache = {}
+
+        self._skip_hash_check = False
+        self._skip_file = {}  # Dict[str, str]
+
+    def require(self, name, urls, sha_hash):
+        """decorator for doing stuff
+
+        Parameters
+        ----------
+        name: str
+        The name to reference the file with
+        urls: list
+        List of urls to download the file from
+        sha_hash: str
+        Hash of file
+        """
+        def decorator(func):
+            @functools.wraps(func)
+            def wrapper(*args, **kwargs):
+                # TODO: Refractor into function(s)
+                replace = self._skip_file.get(name, None)
+                if replace:
+                    if replace.startswith('file://'):
+                        file_path = replace[len('file://'):]
+                    else:
+                        file_path = self._cache.download([replace], redownload=True)
+                elif self._skip_hash_check:
+                    # XXX: Should this redownload every time
+                    file_path = self._cache.download(urls, redownload=True)
+                else:
+                    details = self._cache.get_by_hash(sha_hash)
+                    # breakpoint()
+                    # TODO: Better error and error message
+                    err = KeyError("Hash does not match")
+                    if not details:
+                        # In case we are matching by hash and file does not exist
+                        # That might mean the wrong hash is supplied to decorator
+                        # We match by urls to make sure that is not the case
+                        # XXX: Is this required? Would be useful if users want to
```

Comment:
I have no issue with this but I think we would need to raise a warning about it using the url to check over the hash.